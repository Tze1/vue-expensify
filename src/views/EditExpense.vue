<template>
  <div class="editexpense component">
    <h2 class="editexpense-title">Edit Expense</h2>
    <ExpenseForm
      :expense="expense"
      :submitExpense="editExpense"
    />
    <div class="editexpense-actions">
      <button class="editexpense-cancel button exit" @click="onCancel">Cancel</button>&nbsp;
      <button class="editexpense-remove button danger" @click="onRemove">Remove Expense</button>
    </div>
  </div>
</template>

<script>
  import ExpenseForm from '../components/ExpenseForm';

  export default {
    name: 'EditExpense',
    computed: {
      expense () {
        return this.$store.state.expenses.find(
          expense => expense.id === this.$route.params.id
        )
      },
    },
    methods: {
      editExpense (updates) {
        this.$store.dispatch('editExpense', {
          id: this.$route.params.id,
          updates
        });
        this.$router.push('/dashboard');
      },
      onRemove () {
        this.$store.dispatch('removeExpense', this.$route.params.id);
        this.$router.push('/dashboard');
      },
      onCancel () {
        this.$router.replace('/dashboard');
      }
    },
    components: {
      ExpenseForm,
    },
  };
</script>

<style scoped lang="scss">
  @import '../styles/app.scss';
</style>
