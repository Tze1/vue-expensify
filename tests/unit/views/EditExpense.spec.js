import { shallowMount, mount, createLocalVue } from '@vue/test-utils';
import Vue from 'vue';
import Vuex from 'vuex';
import EditExpense from '@/views/EditExpense';
import testExpenses from '../../fixtures/testExpenses';

/* Even though vue-test-utils docs says
* Vue.nextTick is not needed to wait for DOM updates,
* it's needed below for EditExpense method-calls that in turn call
* bootstrap-vue-components' show/hide methods.
*/

const localVue = createLocalVue();
localVue.use(Vuex);

describe('EditExpense.vue', () => {
  const dispatchStub = jest.fn();
  const pushStub = jest.fn();
  const $store = {
    state: {
      expenses: testExpenses,
    },
    dispatch: dispatchStub,
  };
  const $router = {
    push: pushStub,
  };
  const $route = {
    params: { id: '1' },
  };

  it('renders EditExpense view properly', () => {
    const wrapper = mount(EditExpense, {
      mocks: {
        $store,
        $router,
        $route,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });

  it('dispatches editExpense action properly', () => {
    const id = $route.params.id;  // '1' (testExpenses[0])
    const updates = {
      createdAt: testExpenses[0].createdAt,
      description: 'updated description',
      amount: testExpenses[0].amount,
      note: 'updated note',
    };
    const wrapper = shallowMount(EditExpense, {
      localVue,
      mocks: {
        $store,
        $router,
        $route,
      }
    });

    wrapper.vm.editExpense(updates);

    expect(dispatchStub).toHaveBeenCalledWith('editExpense', { id, updates });
    expect(pushStub).toHaveBeenCalledWith('/dashboard');
  });

  it('displays confirm-modal on remove-button click', () => {
    const wrapper = mount(EditExpense, {
      mocks: {
        $store,
        $router,
        $route,
      },
    });

    wrapper.vm.$refs.confirmRemoveModal.hide();

    Vue.nextTick(() => {
      wrapper.find('.editexpense-remove').trigger('click');
    });

    Vue.nextTick(() => {
      // modal should no longer have style="display:none;" once displayed.
      expect(wrapper.find('.editexpense-confirm .modal').
        attributes('style')).toMatch('');
      wrapper.vm.$refs.confirmRemoveModal.hide();
    });

  });

  it('hides confirm-modal on modal-cancel-button click', () => {
    const wrapper = mount(EditExpense, {
      mocks: {
        $store,
        $router,
        $route,
      },
    });

    wrapper.vm.$refs.confirmRemoveModal.show();
    // modal should no longer have style="display:none;" once displayed.

    Vue.nextTick(() => {
      wrapper.find('.modal-footer .btn-secondary').trigger('click');
      // modal should now have style="display:none;" once displayed.
    });

    Vue.nextTick(() => {
      expect(wrapper.find('.editexpense-confirm .modal').
        attributes('style')).toMatch('display: none');
    });
  });

  it('hides confirm-modal & dispatches removeExpense action properly on modal-remove-button click', () => {
    const id = $route.params.id;
    const wrapper = mount(EditExpense, {
      localVue,
      mocks: {
        $store,
        $router,
        $route
      }
    });

    wrapper.vm.onRemove();  // displays confirm-modal.

    Vue.nextTick(() => {
      wrapper.find('.modal-footer .btn-danger').trigger('click');
      // confirm-modal should now be hidden.
    });

    Vue.nextTick(() => {
      expect(wrapper.find('.editexpense-confirm .modal').
        attributes('style')).toMatch('display: none');
      expect(dispatchStub).toHaveBeenCalledWith('removeExpense', id);
    });
  });
});
