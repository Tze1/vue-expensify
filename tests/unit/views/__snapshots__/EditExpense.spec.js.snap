// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpense.vue renders EditExpense view properly 1`] = `
<div class="editexpense component">
  <h2 class="editexpense-title">Edit Expense</h2>
  <expenseform-stub submitexpense="function () { [native code] }" expense="[object Object]"></expenseform-stub>
  <div class="editexpense-actions"><button class="editexpense-cancel button exit">Cancel</button>&nbsp;
    <button class="editexpense-remove button danger">Remove Expense</button></div>
</div>
`;
